axes/handlers/cache.py

```diff
@@ -0,0 +1,116 @@
+from logging import getLogger
+
+from axes.conf import settings
+from axes.exceptions import AxesSignalPermissionDenied
+from axes.handlers.base import AxesBaseHandler
+from axes.signals import user_locked_out
+from axes.utils import (
+    get_axes_cache,
+    get_client_cache_key,
+    get_client_ip_address,
+    get_client_path_info,
+    get_client_str,
+    get_client_username,
+    get_client_user_agent,
+    get_credentials,
+    get_cool_off,
+)
+
+log = getLogger(settings.AXES_LOGGER)
+
+
+class AxesCacheHandler(AxesBaseHandler):  # pylint: disable=too-many-locals
+    """
+    Signal handler implementation that records user login attempts to cache and locks users out if necessary.
+    """
+
+    def get_failures(self, request, credentials=None, attempt_time=None) -> int:
+        cache = get_axes_cache()
+        cache_key = get_client_cache_key(request, credentials)
+        return cache.get(cache_key, default=0)
+
+    def user_login_failed(self, sender, credentials, request=None, **kwargs):  # pylint: disable=too-many-locals
+        """
+        When user login fails, save attempt record in cache and lock user out if necessary.
+
+        :raises AxesSignalPermissionDenied: if user should be locked out.
+        """
+
+        if request is None:
+            log.error('AXES: AxesCacheHandler.user_login_failed does not function without a request.')
+            return
+
+        username = get_client_username(request, credentials)
+        ip_address = get_client_ip_address(request)
+        user_agent = get_client_user_agent(request)
+        path_info = get_client_path_info(request)
+        client_str = get_client_str(username, ip_address, user_agent, path_info)
+
+        if self.is_whitelisted(request, credentials):
+            log.info('AXES: Login failed from whitelisted client %s.', client_str)
+            return
+
+        failures_since_start = 1 + self.get_failures(request, credentials)
+
+        if failures_since_start > 1:
+            log.warning(
+                'AXES: Repeated login failure by %s. Count = %d of %d. Updating existing record in the cache.',
+                client_str,
+                failures_since_start,
+                settings.AXES_FAILURE_LIMIT,
+            )
+        else:
+            log.warning(
+                'AXES: New login failure by %s. Creating new record in the cache.',
+                client_str,
+            )
+
+        cache = get_axes_cache()
+        cache_key = get_client_cache_key(request, credentials)
+        cache_timeout = get_cool_off().total_seconds()
+
+        cache.set(cache_key, failures_since_start, cache_timeout)
+
+        if failures_since_start >= settings.AXES_FAILURE_LIMIT:
```

Comment:
Since the `failures_since_start` logic seems to apply to both handlers, you could move it into a class method so it wouldn't need to be written out unless a handler wanted to override it.

Reply:
Yeah, these checks should ideally be refactored so that they are in common places in the base implementation.